common/uid_store.c

```diff
@@ -17,22 +17,72 @@
  * Copyright (C) 2015 Simon Newton
  */
 
-#include "uid_store.h"
 #include "common_settings.h"
+#include "config_options.h"
+#include "uid_store.h"
 
 #include <stdint.h>
 
-#ifdef CFG_USE_DEVELOPMENT_UID
-// Fallback to a static development UID
-static const uint8_t test_uid[UID_LENGTH] = {
-  0x7a, 0x70, 0xff, 0xff, 0xff, 0x00
+#ifndef CFG_UID_SOURCE
+#error "CFG_UID_SOURCE is not set"
+#endif
+
+#if CFG_UID_SOURCE == UID_FROM_MAC
+#include "peripheral/eth/plib_eth.h"
+
+static uint8_t kUIDArray[UID_LENGTH];
+
+inline uint8_t ShiftRight(uint8_t b) {
+  return (b >> 4) & 0x0f;
+}
+
+inline uint8_t ShiftLeft(uint8_t b) {
+  return (b << 4) & 0xf0;
+}
+
+void UIDStore_Init() {
+  // The UID is derived from the RDM manufacturer ID & the MAC address.
+  // The first 3 bytes of the MAC address is the Microchip OID 00:1e:c0 which
+  // is constant. The bottom 3 bytes contain the unique serial number.
+  //
+  // To support more than one responder per device, we set the lower 4 bits of
+  // the UID to 0 so we have 16 responders per device. These means the complete
+  // UID takes the form:
+  //   MMMM:0AAAAAA0
```

Comment:
I still think we should keep a track on the wiki from a project perspective, so we know what's being used where; likewise changing the first digit to a 1 helps us as a project. Perhaps we should also add a note about usage on that wiki page. If we do get abuse FTDI clone style, it would be better if they don't clash with our existing UID allocations anyway.